File: src/ReplicatedStorage/Scripts/Scorer.server.lua
local part = script.Parent
local utils = require(game.ServerScriptService.Utils)
local pinball = utils.getParentPinball(part)

part.Touched:Connect(function(hit)
	part.Material = Enum.Material.Neon
	local player = utils.getPinballOwner(pinball)
	if player ~= nil then
		utils.addScore(player, 1000)
	end
	local audio = pinball:FindFirstChild("Audio")
	if audio ~= nil then
		utils.playSound(part, audio.ScorerSound.SoundId)
	end
	task.wait(0.5)
	part.Material = Enum.Material.Plastic
end)

File: src/ServerScriptService/components/Scorer.server.ts
import { BaseComponent, Component } from '@flamework/components'
import { OnStart } from '@flamework/core'
import { store } from 'ServerScriptService/store'
import {
  getArcadeTableOwner,
  getParentArcadeTable,
  playSound,
} from 'ServerScriptService/utils'

@Component({ tag: 'Scorer' })
export class MyComponent extends BaseComponent implements OnStart {
  onStart() {
    const part = this.instance as BasePart
    const arcadeTable = getParentArcadeTable(this.instance)

    part.Touched?.Connect((hit) => {
      part.Material = Enum.Material.Neon

      const player = getArcadeTableOwner(arcadeTable)
      if (player) store.addScore(player.UserId, 1000)

      const audio = arcadeTable.FindFirstChild('Audio') as
        | { ScorerSound?: Sound }
        | undefined
      if (audio?.ScorerSound) playSound(part, audio.ScorerSound.SoundId)

      task.wait(0.5)
      part.Material = Enum.Material.Plastic
    })
  }
}
